### PR TITLE
BigDec operation speedups

### PR DESCRIPTION
**Pull request by ValarDragon**
Changes:
* Rename precisionReuse -> defaultBigDecPrecisionReuse
* Rename precisionReuseSDK -> precisionReuseSDKDec
* Use squaredPrecisionReuse properly (saving many internal ops). This is clearly state compat because we are replacing double Mul's with a single one that will commute.
* expose precisionReuse as an argument more consistently
* Remove redundant copies in many BigDec operations
* Speedup common BigDec X Dec ops
  * Just needs tests
* Speedup Quo w/ rounding - Update: My intuition was wrong for roundup, Now looking into RoundDown
  * Truncate was significantly sped up. (Saved the chop routine, and operate on something two word sizes smaller!)

Its easiest to review this PR commit by commit

### CODE REVIEW

**Comment by ValarDragon** [comment]
I'll edit several CL methods in a second PR

File: osmomath/decimal.go [hunk 1]
```diff
@@ -338,10 +350,36 @@ func (d BigDec) MulMut(d2 BigDec) BigDec {
...
+	return copy
+}
+
+func (d BigDec) MulDecMut(d2 Dec) BigDec {
+	d.i.Mul(d.i, d2.BigIntMut())
+	d.i = chopPrecisionAndRoundSdkDec(d.i)
+
+	if d.i.BitLen() > maxDecBitLen {
+		panic("Int overflow")
+	}
```
Comment (czarcas7ic) [inline osmomath/decimal.go:365]: nit: this logic gets reused 5/6 times in this PR, maybe could be a single line helper.

**Review by czarcas7ic — APPROVED** [review]
I think the go.mod at the highest level needs to be updated to use this version of osmomath in order to ensure all tests pass, and not just the osmomath specific tests (it LGTM, but would make sense to do this prior to merge)

File: osmomath/decimal.go (hunk 1 above)
Comment (ValarDragon) [reply osmomath/decimal.go:365]: Lets do that in a second PR, agreed tho

**Pull request merged**